resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/KubernetesClusterSchedulerBackend.scala

```diff
@@ -0,0 +1,440 @@
+          logDebug("Waiting for pending executors before scaling")
+        } else if (totalExpectedExecutors.get() <= runningExecutorsToPods.size) {
+          logDebug("Maximum allowed executor limit reached. Not scaling up further.")
+        } else {
+          val nodeToLocalTaskCount = getNodesWithLocalTaskCounts
+          for (i <- 0 until math.min(
+            totalExpectedExecutors.get - runningExecutorsToPods.size, podAllocationSize)) {
+            val (executorId, pod) = allocateNewExecutorPod(nodeToLocalTaskCount)
+            runningExecutorsToPods.put(executorId, pod)
+            runningPodsToExecutors.put(pod.getMetadata.getName, executorId)
+            logInfo(
+              s"Requesting a new executor, total executors is now ${runningExecutorsToPods.size}")
+          }
+        }
+      }
+    }
+
+    def handleDisconnectedExecutors(): Unit = {
+      // For each disconnected executor, synchronize with the loss reasons that may have been found
+      // by the executor pod watcher. If the loss reason was discovered by the watcher,
+      // inform the parent class with removeExecutor.
+      disconnectedPodsByExecutorIdPendingRemoval.keys().asScala.foreach { case (executorId) =>
+        val executorPod = disconnectedPodsByExecutorIdPendingRemoval.get(executorId)
+        val knownExitReason = Option(podsWithKnownExitReasons.remove(
+          executorPod.getMetadata.getName))
+        knownExitReason.fold {
+          removeExecutorOrIncrementLossReasonCheckCount(executorId)
+        } { executorExited =>
+          logDebug(s"Removing executor $executorId with loss reason " + executorExited.message)
+          removeExecutor(executorId, executorExited)
+          // We keep around executors that have exit conditions caused by the application. This
+          // allows them to be debugged later on. Otherwise, mark them as to be deleted from the
+          // the API server.
+          if (!executorExited.exitCausedByApp) {
+            deleteExecutorFromClusterAndDataStructures(executorId)
+          }
+        }
+      }
+    }
+
+    def removeExecutorOrIncrementLossReasonCheckCount(executorId: String): Unit = {
+      val reasonCheckCount = executorReasonCheckAttemptCounts.getOrElse(executorId, 0)
+      if (reasonCheckCount >= MAX_EXECUTOR_LOST_REASON_CHECKS) {
+        removeExecutor(executorId, SlaveLost("Executor lost for unknown reasons."))
+        deleteExecutorFromClusterAndDataStructures(executorId)
+      } else {
+        executorReasonCheckAttemptCounts.put(executorId, reasonCheckCount + 1)
+      }
+    }
+
+    def deleteExecutorFromClusterAndDataStructures(executorId: String): Unit = {
+      disconnectedPodsByExecutorIdPendingRemoval.remove(executorId)
+      executorReasonCheckAttemptCounts -= executorId
+      RUNNING_EXECUTOR_PODS_LOCK.synchronized {
+        runningExecutorsToPods.remove(executorId).map { pod =>
+          kubernetesClient.pods().delete(pod)
```

Comment:
Can `delete()` block ? If yes, the actual delete could be done outside the sync block ?
(Same pattern elsewhere in the code as well - except for cases where we cant remove)

Reply:
All Kubernetes API calls are asynchronous. We can technically stall if there's an HTTP connection error though.

Reply:
Sounds good.

So in `stop`, if `delete` is async, will the pods be deleted ?
Or does `kubernetesClient.close` flush all requests and block for their processing.

Reply:
By asynchronous I mean that after the HTTP request is sent, the response comes back before the Kubernetes master has actually done the work of tearing the pod down.

All calls are synchronous with respect to the HTTP call, but the work done in response to HTTP requests are themselves done asynchronously. So we could still bottleneck here if e.g. there's a connection timeout to the Kubernetes API. It might not be a significant difference in this case because a breakdown in communication with the API server is going to cause all sorts of problems outside of locked code in the first place.

Reply:
But if we're not losing anything correctness-wise by moving the potentially slow calls out of synchronized blocks then we should do so. I think it's safe to move the code out, but will double check and mention if there are problems with that.